### PR TITLE
Support exclusively positive integer columns

### PR DESCRIPTION
**Pull request by gracexin2003**
Add a parameter to support columns with both positive+negative values and exclusively positive values.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by shixuan-fan — COMMENTED** [review]
It might be a dumb question because I did not see how this is used: what is the motivation behind this change?

**Comment by gracexin2003** [comment]
> It might be a dumb question because I did not see how this is used: what is the motivation behind this change?

Before, my negative integers version assumed that every column would have both negative and positive integers ranging from `[-(1 << encodingBits), (1 << encodingBits) - 1]`. 
However, some columns may have exclusively positive values from [0, (1 << (encodingBits+1)) - 1], and because Presto wants continuous z-order file addresses, the addresses would be wrong in the negative integers version. So this change gives an option of choosing both negative+positive integers, or just positive integers.

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java [hunk 1]
```diff
@@ -42,15 +44,18 @@
...
 
         this.encodingBits = encodingBits;
+        this.onlyPositive = onlyPositive;
 
-        totalBitLength = encodingBits.stream().mapToInt(Integer::intValue).sum() + encodingBits.size();
-        maxBitLength = encodingBits.stream().mapToInt(Integer::intValue).max().getAsInt() + 1;
+        int totalBitLength = encodingBits.stream().mapToInt(Integer::intValue).sum();
+        this.totalBitLength = onlyPositive ? totalBitLength : totalBitLength + encodingBits.size();
+        int maxBitLength = encodingBits.stream().mapToInt(Integer::intValue).max().getAsInt();
+        this.maxBitLength = onlyPositive ? maxBitLength : maxBitLength + 1;
```
Comment (lliiun-z) [inline presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:58]: Can we do something as follows? 

this.totalBitLength = ...;
this.maxBitLength = ...;
if (!onlyPositive) {
    this.totalBitLength += encodingBits.size();
    this.maxBitLength += 1;
}

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java [hunk 2]
```diff
@@ -66,7 +71,7 @@ public ZOrder(List<Integer> encodingBits)
         List<Integer> bitLengths = new ArrayList<>(encodingBits);
         for (int dimensionIndex = maxBitLength - 1; dimensionIndex >= 0; dimensionIndex--) {
             for (int bitLengthIndex = 0; bitLengthIndex < bitLengths.size(); bitLengthIndex++) {
-                if (bitLengths.get(bitLengthIndex) >= 0) {
+                if (onlyPositive ? bitLengths.get(bitLengthIndex) > 0 : bitLengths.get(bitLengthIndex) >= 0) {
```
Comment (lliiun-z) [inline presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:74]: If in Line 71, we check onlyPositive value, and adjust values in bitLengths accordingly, then we don't need to check onlyPositive for so many times. E.g.,

if (!onlyPositive) {
   for (int i = 0; i < bitLengths.size(); ++i) {
      bitLengths.set(i, bitLengths.get(i) + 1);
  }
}

**Comment by lliiun-z** [comment]
@gracexin2003 , overall I think the diff is accurate since I tested it by running a few queries. One big question that I have is: Can we make the modification into as fewer places as possible? This is so called Single Responsibility Principle. What I worry is if we want to support some negative columns and some positive columns, do we need to add another flag? How many places do we need to change? If we can organize the code such that we only need to modify a few places, the code will be easier to read and modify.

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java (hunk 1 above)
Comment (gracexin2003) [reply presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:58]: Both totalBitLength and maxBitLength are final variables, so I can't modify them after their first instantiation. That is why I initially created the local variables to store the positive values.

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java [hunk 3]
```diff
@@ -34,6 +34,8 @@
 
     // Assume encodingBits is 0-based
     private final List<Integer> encodingBits;
+    private final boolean onlyPositive;
```
Comment (highker) [inline presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:37]: nit: `positiveIntegersOnly`

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java [hunk 4]
```diff
@@ -42,15 +44,18 @@
     /**
      * Class constructor specifying the number of bits each value will take up for encoding and decoding.
      */
-    public ZOrder(List<Integer> encodingBits)
+    public ZOrder(List<Integer> encodingBits, boolean onlyPositive)
```
Comment (highker) [inline presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:47]: I would create another constructor with default onlyPositive=false

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java (hunk 2 above)
Comment (highker) [reply presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:74]: +1, I'm not sure if JIT compiler would be smart enough to optimize the looping branch

**Review by highker — APPROVED** [review]
minor comments only; will leave @lliiun-z for logic changes

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java (hunk 2 above)
Comment (lliiun-z) [reply presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:74]: This is a good exercise to understand how JIT works for this case. :)

**Review by shixuan-fan — APPROVED** [review]
LGTM % existing comments :)

**Comment by highker** [comment]
any progress on this PR?

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java [hunk 5]
```diff
@@ -42,15 +44,37 @@
...
+        this.positiveIntegersOnly = false;
 
-        totalBitLength = encodingBits.stream().mapToInt(Integer::intValue).sum() + encodingBits.size();
-        maxBitLength = encodingBits.stream().mapToInt(Integer::intValue).max().getAsInt() + 1;
+        int totalBitLength = encodingBits.stream().mapToInt(Integer::intValue).sum();
+        this.totalBitLength = totalBitLength + encodingBits.size();
+        int maxBitLength = encodingBits.stream().mapToInt(Integer::intValue).max().getAsInt();
+        this.maxBitLength = maxBitLength + 1;
 
         dimensions = initializeDimensions();
```
Comment (highker) [inline presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:79]: not really....
```java
public ZOrder(List<Integer> encodingBits)
{
    this(encodingBits, false);
}

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java [hunk 6]
```diff
@@ -42,15 +44,37 @@
...
+        int maxBitLength = encodingBits.stream().mapToInt(Integer::intValue).max().getAsInt();
+        this.totalBitLength = positiveIntegersOnly ? totalBitLength : totalBitLength + encodingBits.size();
+        this.maxBitLength = positiveIntegersOnly ? maxBitLength : maxBitLength + 1;
+
+        dimensions = initializeDimensions();
+    }
+
+    /**
+     * Class constructor with default <code>positiveIntegersOnly</code> false
+     */
```
Comment (highker) [inline presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:65]: remove this javadoc; it's self-explanatory

File: presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java [hunk 7]
```diff
@@ -64,11 +88,22 @@ public ZOrder(List<Integer> encodingBits)
     {
         int[] dimensions = new int[maxBitLength];
         List<Integer> bitLengths = new ArrayList<>(encodingBits);
+
         for (int dimensionIndex = maxBitLength - 1; dimensionIndex >= 0; dimensionIndex--) {
-            for (int bitLengthIndex = 0; bitLengthIndex < bitLengths.size(); bitLengthIndex++) {
-                if (bitLengths.get(bitLengthIndex) >= 0) {
-                    dimensions[dimensionIndex]++;
-                    bitLengths.set(bitLengthIndex, bitLengths.get(bitLengthIndex) - 1);
+            if (positiveIntegersOnly) {
```
Comment (highker) [inline presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java:93]: I would put `if` outside the for loop for performance reasons

**Pull request merged**